File: leetcode3/이진희/877. Stone Game.java
/*

1. 아이디어 : 짝수개의 돌, 돌의 총 개수는 홀수
              A->B로 번갈아가며 돌을 선택하고 이때 양끝 돌 중 하나를 선택
              
              A가 이기면 true, B가 이기면 false

              순서가 A부터 고정되고 가져가는 돌의 위치도 정해져 있어, 생각해보면 A는 항상 이길 수 있다.


2. 시간복잡도 : O(1)
자료구조/알고리즘 : 아이디어

*/

class Solution {
    public boolean stoneGame(int[] piles) {
        // A -> B
        // 맨 앞과 맨 끝 중 돌 가져가기

        // int l = 0;
        // int r = piles.length-1;
        // int a = 0;
        // int b = 0;

        // while(l<r) {
        //     // A
        //     if(piles[l]>=piles[r]) {
        //         a+=piles[l];
        //         l++;
        //     }
        //     else {
        //         a+=piles[r];
        //         r--;
        //     }

        //     // B
        //     if(piles[l]>piles[r]) {
        //         b+=piles[r];
        //         r--;
        //     }
        //     else {
        //         b+=piles[l];
        //         l++;
        //     }
        // }

        // return a>b;

        return true;
    }
}

File: leetcode3/최원준/877. Stone Game.py
#

'''
1. 아이디어 :
선택할 수 있는 돌의 길이를 1부터 계산.
반복 계산을 하지 않기 위해 dp 사용.
l~r에서 가장 효율적으로 돌을 가져가는 방법은 l + (l+1 ~ r) 또는 (l~r-1) + r

2. 시간복잡도 :
    O(n + n * n)

3. 자료구조/알고리즘 :
dp

'''
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        n = len(piles)

        dp = [[0] * n for _ in range(n)]

        for i in range(n):
            dp[i][i] = piles[i]
        

        for length in range(2, n+1):
            for lp in range(n - length + 1):
                rp = lp + length - 1

                left = piles[lp] - dp[lp+1][rp]
                right = piles[rp] - dp[lp][rp-1]
                dp[lp][rp] = max(left, right)

        return dp[0][n-1] > 0
